llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp

```diff
@@ -409,49 +453,121 @@ static void adjustByValArgAlignment(Argument *Arg, Value *ArgInParamAS,
+    if (*U == II.getRawDest() && !IsGridConstant)
+      PI.setAborted(&II);
+
+    // TODO: memcpy from arg is OK as it can get unrolled into ld.param.
+    // However, memcpys are currently expected to be unrolled before we
+    // get here, so we never see them in practice, and we do not currently
+    // handle them when we convert IR to access param space directly. So,
+    // we'll mark it as an escape for now. It would still force a copy on
+    // pre-sm_70 GPUs where we can't take address of a parameter w/o a copy.
+    //
+    // PI.setEscaped(&II);
+  }
+
+  void visitMemSetInst(MemSetInst &II) {
+    if (*U == II.getRawDest() && !IsGridConstant)
+      PI.setAborted(&II);
+  }
+  // debug only helper.
+  auto &getVisitedUses() { return VisitedUses; }
+};
+} // namespace
 void NVPTXLowerArgs::handleByValParam(const NVPTXTargetMachine &TM,
                                       Argument *Arg) {
-  bool IsGridConstant = isParamGridConstant(*Arg);
   Function *Func = Arg->getParent();
+  bool HasCvtaParam = TM.getSubtargetImpl(*Func)->hasCvtaParam();
+  bool IsGridConstant = HasCvtaParam && isParamGridConstant(*Arg);
+  const DataLayout &DL = Func->getDataLayout();
   BasicBlock::iterator FirstInst = Func->getEntryBlock().begin();
   Type *StructType = Arg->getParamByValType();
   assert(StructType && "Missing byval type");
 
-  auto AreSupportedUsers = [&](Value *Start) {
-    SmallVector<Value *, 16> ValuesToCheck = {Start};
-    auto IsSupportedUse = [IsGridConstant](Value *V) -> bool {
-      if (isa<GetElementPtrInst>(V) || isa<BitCastInst>(V) || isa<LoadInst>(V))
-        return true;
-      // ASC to param space are OK, too -- we'll just strip them.
-      if (auto *ASC = dyn_cast<AddrSpaceCastInst>(V)) {
-        if (ASC->getDestAddressSpace() == ADDRESS_SPACE_PARAM)
-          return true;
-      }
-      // Simple calls and stores are supported for grid_constants
-      // writes to these pointers are undefined behaviour
-      if (IsGridConstant &&
-          (isa<CallInst>(V) || isa<StoreInst>(V) || isa<PtrToIntInst>(V)))
-        return true;
-      return false;
-    };
-
-    while (!ValuesToCheck.empty()) {
-      Value *V = ValuesToCheck.pop_back_val();
-      if (!IsSupportedUse(V)) {
-        LLVM_DEBUG(dbgs() << "Need a "
-                          << (isParamGridConstant(*Arg) ? "cast " : "copy ")
-                          << "of " << *Arg << " because of " << *V << "\n");
-        (void)Arg;
-        return false;
-      }
-      if (!isa<LoadInst>(V) && !isa<CallInst>(V) && !isa<StoreInst>(V) &&
-          !isa<PtrToIntInst>(V))
-        llvm::append_range(ValuesToCheck, V->users());
-    }
-    return true;
-  };
-
-  if (llvm::all_of(Arg->users(), AreSupportedUsers)) {
+  ArgUseChecker AUC(DL, IsGridConstant);
+  ArgUseChecker::PtrInfo PI = AUC.visitArgPtr(*Arg);
+  // Easy case, accessing parameter directly is fine.
+  if (!(PI.isEscaped() || PI.isAborted()) && AUC.Conditionals.empty()) {
```

Comment:
I see that the Conditionals set is used for debug logs, but for pure functionality, having a local variable for the conditionals, and a simple `NoConditionalUses` boolean might make more sense? nbdl, and I'm not sure about this.

Reply:
Yeah. Readability of this is not great. Will fix.

I am currently collecting more info than strictly needed. At the moment we indeed only care whether there are any conditionals or not, but I am considering further improvements.  E.g. if we have `load(select(cond, arg_ptr, non_arg_ptr))` our choice now is to either use `cvta.param` on sm_70+ or force a copy. If we transform it to `select(cond, load(arg_ptr), load(non_arg_ptr))`, then we can avoid the copy on the older GPUs, too. 

Would you happen to know whether there's any overhead associated with using the pointer obtained via `cvta.param` vs direct access to the parameter via `ld.param` ?

Reply:
This is something that I'm not sure about, hence https://github.com/llvm/llvm-project/pull/106423#issuecomment-2316126096

Reply:
The overhead is similar to generic loads vs non-generic loads. Generic loads via cvta.param are slower because it requires runtime address space check and conversion. 

> . If we transform it to select(cond, load(arg_ptr), load(non_arg_ptr)), then we can avoid the copy on the older GPUs, too.

This kind of speculative load may be illegal.

Reply:
Yeah, there's that, but we should be able to create correct equivalent w/o speculatively loading stuff, with conditional branches. In any case, all these future improvements ideas are just hand waving at the moment.

I'll test the changes on a large set of compilations. If nothing interesting pops up, I can  trim down collection of data that's not strictly necessary. Or put it under `LLVM_DEBUG`.

Reply:
Yeah, there's that, but we should be able to create correct equivalent w/o speculatively loading stuff, with conditional branches. In any case, all these future improvements ideas are just hand waving at the moment.

I'll test the changes on a large set of compilations. If nothing interesting pops up, I can  trim down collection of data that's not strictly necessary. Or put it under `LLVM_DEBUG`.